Add outputs for strain and stress

einola opened the pull request:
# Add outputs for strain and stress

Fixes #692 

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

This PR provides shear, divergence, sigma_I and sigma_II as outputs through ConfigOutput. The code already contained routines to calculate these values, but they were not being passed to the output mechanism. Shear is output as 1/s, and not as log10(1/s) as in the original code.

---
# Test Description

The methods used to calculate the values were already tested, so testing the new code simply consisted of checking that the output looked reasonably similar to that provided by the old VTK routines.

---
# Documentation Impact

None

---
# Other Details

None
---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

einola committed:
Output shear, divergence, sigmaI, and sigmaII

The routines to calculate those already existed, so it's mostly a
question of plumbing.

einola committed:
Shear should be in 1/s not log10(1/s)

The user can change the units afterwards.

einola committed:
Output shear by default in the benchmark

Because it's what whe look at the most

timspainNERSC commented on core/src/include/ModelComponent.hpp:
```diff
@@ -74,6 +74,10 @@ namespace Protected {
     inline constexpr TextTag SLAB_FDW
         = "SLAB_FDW"; // Slab ocean salinity nudging water flux, kg s⁻¹ m⁻²
+    inline constexpr TextTag SHEAR = "SHEAR"; // Instantaneous share rate, s⁻¹
+    inline constexpr TextTag DIV = "DIV"; // Instantaneous divergnece , s⁻¹
```
"divergence"

timspainNERSC commented on core/src/modules/DynamicsModule/MEVPDynamics.cpp:
```diff
@@ -74,6 +74,11 @@ MEVPDynamics::MEVPDynamics()
 {
     getStore().registerArray(Protected::ICE_U, &uice, RO);
     getStore().registerArray(Protected::ICE_V, &vice, RO);
+
+    getStore().registerArray(Protected::SHEAR, &shear, RO);
+    getStore().registerArray(Protected::DIV, &divergence, RO);
+    getStore().registerArray(Protected::SIGMAI, &sigmaI, RO);
+    getStore().registerArray(Protected::SIGMAII, &sigmaII, RO);
```
Move to `IDynamics`.

einola replied:
A very good point. I also took the liberty of moving the registration of uice and vice to IDynamics.

timspainNERSC commented on core/src/modules/DynamicsModule/BBMDynamics.cpp:
```diff
@@ -92,6 +92,11 @@ BBMDynamics::BBMDynamics()
 {
     getStore().registerArray(Protected::ICE_U, &uice, RO);
     getStore().registerArray(Protected::ICE_V, &vice, RO);
+
+    getStore().registerArray(Protected::SHEAR, &shear, RO);
+    getStore().registerArray(Protected::DIV, &divergence, RO);
+    getStore().registerArray(Protected::SIGMAI, &sigmaI, RO);
+    getStore().registerArray(Protected::SIGMAII, &sigmaII, RO);
```
Move to `IDynamics`.

timspainNERSC commented on run/config_benchmark.cfg:
```diff
@@ -15,5 +15,5 @@ OceanBoundaryModule = Nextsim::BenchmarkOcean
 period = P0-0T04:00:00
 start = 2010-01-01T12:00:00Z
-field_names = hsnow,hice,cice,u,v
+field_names = hsnow,hice,cice,u,v,shear
```
If adding shear here, it seems like some kind of test of the values should be present in `test_integration_test.py`.

einola replied:
This is in the cyclone benchmark config file, so it's not related to the integration test. We should consider making this benchmark an additional integration test, but that's a separate issue.

timspainNERSC replied:
Oh, of course it is.

timspainNERSC reviewed (changes requested):
Since the added arrays are owned by `IDynamics`, they should be registered when setting up that class (as `taux`/`y` are). Also, if you are outputting the shear in the integration test, the values should be at least cursorily checked.

einola committed:
Moving common dynamics variables to IDynamics.hpp

We now register divergence, u, v, taux, tauy, shear, sigmaI, sigmaII,
and damage all in the IDynamics base class and not in the child classes.

einola committed:
Merge branch 'develop' into issue692_output_strain_stress

# Conflicts:
#	core/src/include/ModelComponent.hpp

tdcwilliams commented:
Hi @einola - I noticed that `params.expPmax` never got changed to `params.expPmax + 1` in:
https://github.com/nextsimhub/nextsimdg/blob/23791d16548e2229bdb2c8db8d4933484c189c69/dynamics/src/include/BBMStressUpdateStep.hpp#L90C1-L91C81

einola commented:
> Hi @einola - I noticed that `params.expPmax` never got changed to `params.expPmax + 1` in: https://github.com/nextsimhub/nextsimdg/blob/23791d16548e2229bdb2c8db8d4933484c189c69/dynamics/src/include/BBMStressUpdateStep.hpp#L90C1-L91C81

Hi @tdcwilliams, Yes, that fix somehow got lost again. Can you write an issue and create a PR, and then I can approve it?
